fvm/environment/random_generator.go

```diff
@@ -0,0 +1,142 @@
+package environment
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/onflow/flow-go/crypto/random"
+	"github.com/onflow/flow-go/fvm/errors"
+	"github.com/onflow/flow-go/fvm/storage/state"
+	"github.com/onflow/flow-go/fvm/tracing"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/trace"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/state/protocol/prg"
+)
+
+type RandomGenerator interface {
+	// UnsafeRandom returns a random uint64
+	// Todo: rename to Random() once Cadence interface is updated
+	UnsafeRandom() (uint64, error)
+}
+
+var _ RandomGenerator = (*randomGenerator)(nil)
+
+// randomGenerator implements RandomGenerator and is used
+// for the transactions execution environment
+type randomGenerator struct {
+	tracer tracing.TracerSpan
+
+	stateSnapshot protocol.Snapshot
+	txId          flow.Identifier
+
+	prg          random.Rand
+	isPRGCreated bool
+}
+
+type ParseRestrictedRandomGenerator struct {
+	txnState state.NestedTransactionPreparer
+	impl     RandomGenerator
+}
+
+func NewParseRestrictedRandomGenerator(
+	txnState state.NestedTransactionPreparer,
+	impl RandomGenerator,
+) RandomGenerator {
+	return ParseRestrictedRandomGenerator{
+		txnState: txnState,
+		impl:     impl,
+	}
+}
+
+func (gen ParseRestrictedRandomGenerator) UnsafeRandom() (
+	uint64,
+	error,
+) {
+	return parseRestrict1Ret(
+		gen.txnState,
+		trace.FVMEnvRandom,
+		gen.impl.UnsafeRandom)
+}
+
+func NewRandomGenerator(
+	tracer tracing.TracerSpan,
+	stateSnapshot protocol.Snapshot,
+	txId flow.Identifier,
+) RandomGenerator {
+	gen := &randomGenerator{
+		tracer:        tracer,
+		stateSnapshot: stateSnapshot,
+		txId:          txId,
+		isPRGCreated:  false, // PRG is not created
+	}
+
+	return gen
+}
+
+func (gen *randomGenerator) createPRG() (random.Rand, error) {
+	// Use the protocol state source of randomness [SoR] for the current block's
+	// execution
+	source, err := gen.stateSnapshot.RandomSource()
```

Comment:
That is a big interface to be dragging around when we only need one function. 

Would it make sense to create a 
```
type RandomSourceProvider interface {
	RandomSource() (SomethingSomething, error)
}
```
in the FVM, so we don't have to use the whole snapshot. This will make it easier to test too.

Reply:
I agree the interface is large while we only require one function. I started to refactor by adding another interface `RandomSourceProvider` or `EntropyProvider` but I noticed the code became a bit more complex. 

You can notice that `RandomSource` isn't abstracted from the protocol. In particular, the [description and errors ](https://github.com/onflow/flow-go/blob/master/state/protocol/snapshot.go#L109-L121)returned are tight to how the protocol works (this shouldn't be the way to define interfaces IMO but that's how many interfaces are written in our interface). If we will add an interface in FVM for `RandomSource`, we would need to copy the same description and errors from the original `protocol.State` interface. That would make very little sense in the context of FVM and will end being a "fake" abstraction. 
Moreover, we would need to explain why the call [wouldn't lead to errors in normal operations](https://github.com/onflow/flow-go/blob/tarak/env-safe-random/fvm/environment/random_generator.go#L81-L85). The explanation is tight to the protocol state too, and is not abstracted from it (the source is taken from the child block, and it turns out the child of an executable block always exists by the rules of the protocol state). 

It would have been easier if the protocol state interface (and the `RandomSource`) was written in a real abstracted way. Maybe I didn't find the best way to make the code better.

Reply:
I  see.

I still think its worth looking into somehow restricting what is injected into the FVM (in a  future PR), so that the FVM can be instantiated without knowing about Snapshots and ProtocolState. 

But we can do that as a completely separate PR.

Reply:
I thought more about it and made a refactor here : https://github.com/onflow/flow-go/pull/4498/commits/f2bc373a52912c7a936bee6e78e08246fab32add 

To address the awkward misalignment of returned errors, I added comments to explain why we can use `snapshot` (that can error) as the new interface (that shouldn't error), every time we cast `snapshot` as the interface.